types/dropzone/index.d.ts

```diff
@@ -148,6 +148,18 @@ declare namespace Dropzone {
 
         previewTemplate?: string;
     }
+
+    export interface DropzoneListeners {
+        element: HTMLElement;
+        events: {
+            dragstart: (e: DragEvent) => Dropzone;
```

Comment:
I don't think the keys are static in `listeners`
See
https://gitlab.com/meno/dropzone/-/blob/6cef0c303a472468f8d3581c82062a6d31045666/src/dropzone.js#L1195

Also never mentioned in doc
There is already a way to add event listeners
https://www.dropzonejs.com/#events

Reply:
You're right, they are not! My bad. How can I still type this in a way that makes sense? I guess any kind of callback that takes an event, not necessarily a `DragEvent`.

Reply:
> Also never mentioned in doc

It is not, but we make use of it in our code base. For better or worse, there's code like this (simplified):

```JavaScript
dropzone.listeners[0].element = dropzone.element
```

Reply:
```typescript
{
  dragstart?: (e: DragEvent) => Dropzone;
  [index:string]: (e: Event) => Dropzone;
}
```

Use Index Signature
- https://basarat.gitbook.io/typescript/type-system/index-signatures
- https://www.typescriptlang.org/docs/handbook/advanced-types.html#index-types-and-index-signatures

Special already known cases for events you put can on the top

Remember to add `click`

Reply:
> Remember to add `click`

Is this correct?

```TypeScript
click?: (e: MouseEvent) => any;
```

> Use Index Signature

I see, thank you! However I couldn't get it to work the way you described.

```
Property 'dragstart' of type '(e: DragEvent) => Dropzone' is not assignable to string index type '(e: Event) => Dropzone'.ts
```

Is the way to go here using an intersection type like [described here](https://stackoverflow.com/a/45261035/1928168)?

Reply:
If I understood the docs correctly, especially

> All members must conform to the string index signature

I can only do `dragstart?: (e: Event) => Dropzone;`, loosening the type a bit from `DragEvent` to `Event` otherwise.

Reply:
This is the best I can come up:
```typescript
interface DropzoneListener {
  element: HTMLElement;
  events: {
    dragstart?:   (e: DragEvent) => Dropzone;
    dragenter?:   (e: DragEvent) => Dropzone;
    dragover?:    (e: DragEvent) => Dropzone;
    dragleave?:   (e: DragEvent) => Dropzone;
    drop?:        (e: DragEvent) => Dropzone;
    dragend?:     (e: DragEvent) => Dropzone;
    click?:       (e: MouseEvent) => any;

    [key: string]: ((e: any) => any) | undefined;
  };
}
```
I can't put `((e: Event) => any) | undefined` for some reason
Also I think you should remove the `s` in interface name
`DropzoneListener` not `DropzoneListeners`

`listeners: Dropzone.DropzoneListeners[];` feels strange~

Reply:
Good point about the `s` in the name.

I wonder if it's better to have some predefined listeners and not having stronger typings around anything that isn't predefined or just going with something like:

```TypeScript
interface DropzoneListener {
  element: HTMLElement;
  events: {
    [key: string]: ((e: Event) => any) | undefined;
  };
}
```

WDYT?

Reply:
If you don't need the more specific event types
I am fine with it
Do note in that case you should have
`[key: string]: ((e: Event) => any);` without `undefined`
It was added to fix the validation error

Reply:
Thank you for your help! I've pushed the changes and hope this concludes work for this PR.

If there's anything else, let me know!